common/kvstore/src/main/java/org/apache/spark/util/kvstore/InMemoryStore.java

```diff
@@ -322,15 +374,32 @@ private static Object indexValueForEntity(KVTypeInfo.Accessor getter, Object ent
      */
     private List<T> copyElements() {
       if (parent != null) {
-        KVTypeInfo.Accessor parentGetter = ti.getParentAccessor(index);
-        Preconditions.checkArgument(parentGetter != null, "Parent filter for non-child index.");
-        Comparable<?> parentKey = asKey(parent);
-
-        return elements.stream()
-          .filter(e -> compare(e, parentGetter, parentKey) == 0)
-          .collect(Collectors.toList());
+        Comparable<Object> parentKey = asKey(parent);
+        if (!naturalParentIndexName.isEmpty() &&
+          naturalParentIndexName.equals(ti.getParentIndexName(index))) {
```

Comment:
Is it possible that `naturalParentIndexName` doesn't equal to `ti.getParentIndexName(index)`? Isn't `String index = KVIndex.NATURAL_INDEX_NAME`?

Reply:
It is possible.  I have explained in https://github.com/apache/spark/pull/27716/files#r385846069.